plugins/pcgc_cax/pcgc_cax.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"autonity-oracle/plugins/common"
+	"autonity-oracle/types"
+	"encoding/json"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-plugin"
+	"github.com/shopspring/decimal"
+	"io"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+)
+
+// This plugin is only used for autonity round 4 game purpose, the data of NTN-USD & ATN-USD come from a simulated
+// exchange service build by Clearmatics.
+const (
+	version   = "v0.0.1"
+	orderbook = "orderbooks"
+	quote     = "quote"
+	NTNATN    = "NTN-ATN"
+	NTNUSD    = "NTN-USD"
+	ATNUSD    = "ATN-USD"
+)
+
+var defaultConfig = types.PluginConfig{
+	Key:                "",
+	Scheme:             "https",
+	Endpoint:           "cax.devnet.clearmatics.network", // todo: replace it with cax.piccadilly.autonity.org for piccadilly network.
+	Timeout:            10,                               //10s
+	DataUpdateInterval: 3600,                             //3600s, todo: if there is a rate limit for the CAX service, we'd need to set it properly.
+}
+
+type CAXQuote struct {
+	Timestamp string `json:"timestamp"`
+	BidPrice  string `json:"bid_price"`
+	BidAmount string `json:"bid_amount"`
+	AskPrice  string `json:"ask_price"`
+	AskAmount string `json:"ask_amount"`
+}
+
+type CAXClient struct {
+	conf   *types.PluginConfig
+	client *common.Client
+	logger hclog.Logger
+}
+
+func NewCAXClient(conf *types.PluginConfig) *CAXClient {
+	client := common.NewClient(conf.Key, time.Second*time.Duration(conf.Timeout), conf.Endpoint)
+	if client == nil {
+		panic("cannot create client for open exchange rate api")
+	}
+
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name:   "AutonityR4CAX",
+		Level:  hclog.Debug,
+		Output: os.Stdout,
+	})
+
+	return &CAXClient{
+		conf:   conf,
+		client: client,
+		logger: logger,
+	}
+}
+
+func (cc *CAXClient) FetchPrice(symbols []string) (common.Prices, error) {
+	var prices common.Prices
+	priceMap := make(map[string]common.Price)
+
+	for _, s := range symbols {
+		p, err := cc.fetchPrice(s)
+		if err != nil {
+			cc.logger.Error("query price", "error", err.Error())
+			continue
+		}
+		priceMap[s] = p
+		prices = append(prices, p)
+	}
+
+	if len(prices) == 0 {
+		return nil, common.ErrDataNotAvailable
+	}
+
+	// for autonity round4 game, the price of "NTN-ATN" is derived from the price of "NTN-USD" and "ATN-USD"
+	if _, ok := priceMap[NTNATN]; !ok && len(priceMap) == 2 {
+		// since only 3 symbols are supported, thus we assume the collected two symbols are NTN-USD and ATN-USD.
```

Comment:
Maybe I am missing something because I do not know how the CAX works, but isn't this assumption dangerous?

What if we actually fetched NTNATN and ATNUSD for example? Can it happen?

Reply:
At the moment it is only these 2 markets.  But @lorenzo-dev1 is right - theoretically, additional markets could be created on the CAX.

Reply:
As each plugin states there available/supported symbols on its loading phase, thus we always filter out un-supported symbols before we can come to here, this part was implemented as a common logic for plugins, so the fetch() interface of a new plugin will always focus on the fetching the supported symbols from data source.

Reply:
I understand the logic.

My doubt is more about the fact that if by chance one between `priceMap[NTNUSD]` and `priceMap[ATNUSD]` the code will crash in `computeDerivedPrice` as there is no check for nil. However I am not sure if that situation can happen.

Reply:
No, it does not happen at all, since the CAX service only provide 2 symbol pairs: NTN-USD, ATN-USD. However I can add this checking for higher robust in case there is any update in CAX side.

Reply:
Thanks Jason. I think in case one of those map is empty we can still `return prices, nil` (or at least that's what we do if `computeDerivedPrice` returns an err).

Reply:
> Thanks Jason. I think in case one of those map is empty we can still `return prices, nil` (or at least that's what we do if `computeDerivedPrice` returns an err). We won't be able to compute NTNATN but we might have NTNUSD or ATNUSD (it could be that only one is missing).
> 
> But again I am not sure about the likelihood of this happening, so I leave the decision to you @Jason-Zhangxin-Chen

Thanks Lorenzo, I keep the partial data returned via a9eb620860a6be9b0f5b2a6c7b728ec56b966db8.